### PR TITLE
[9.x] Add a new function `withSoftDeletes` in routes

### PR DESCRIPTION
**Pull request by michaelnabil230**
Laravel allows models to have soft delete to not remove from the row form database. to add the date the resource is deleted.
This contribution allows a developer to create a resource route that supports soft delete routes.

For ex:
```php
Route::resource('users', UserController::class)->withSoftDeletes();
```

### CODE REVIEW

**Comment by Rizky92** [comment]
`withTrashed()` already did this.

**Comment by michaelnabil230** [comment]
> `withTrashed()` already did this.

But don't generate routes for trashed or restore methods in your controller

File: src/Illuminate/Routing/ResourceRegistrar.php [hunk 1]
```diff
@@ -92,6 +101,10 @@ public function register($name, $controller, array $options = [])
 
         $defaults = $this->resourceDefaults;
 
+        if (isset($options['withSoftDeletes']) && $options['withSoftDeletes'] == true) {
```
Comment (browner12) [inline src/Illuminate/Routing/ResourceRegistrar.php:104]: use strict equivalency

File: src/Illuminate/Routing/ResourceRegistrar.php [hunk 2]
```diff
@@ -311,6 +324,42 @@ protected function addResourceDestroy($name, $base, $controller, $options)
...
+     * Add the trashed method for a resourceful route.
+     *
+     * @param  string  $name
+     * @param  string  $base
+     * @param  string  $controller
+     * @param  array  $options
+     * @return \Illuminate\Routing\Route
+     */
+    protected function addResourceTrashed($name, $base, $controller, $options)
+    {
```
Comment (browner12) [inline src/Illuminate/Routing/ResourceRegistrar.php:337]: is shallowness relevant here?

File: src/Illuminate/Routing/ResourceRegistrar.php [hunk 3]
```diff
@@ -311,6 +324,42 @@ protected function addResourceDestroy($name, $base, $controller, $options)
...
+     * Add the restore method for a resourceful route.
+     *
+     * @param  string  $name
+     * @param  string  $base
+     * @param  string  $controller
+     * @param  array  $options
+     * @return \Illuminate\Routing\Route
+     */
+    protected function addResourceRestore($name, $base, $controller, $options)
+    {
```
Comment (browner12) [inline src/Illuminate/Routing/ResourceRegistrar.php:355]: is shallowness relevant here?

File: src/Illuminate/Routing/Router.php [hunk 4]
```diff
@@ -359,6 +359,10 @@ public function apiResource($name, $controller, array $options = [])
     {
         $only = ['index', 'show', 'store', 'update', 'destroy'];
 
+        if (isset($options['withSoftDeletes']) && $options['withSoftDeletes'] == true) {
```
Comment (browner12) [inline src/Illuminate/Routing/Router.php:362]: use strict equivalency

**Comment by browner12** [comment]
@Rizky92 `withTrashed()` is similar but different that what is proposed here. `withTrashed()` will load soft deleted models when "route model binding" is happening.

Similar relevant PR I created a couple days ago #44405

**Comment by michaelnabil230** [comment]
A new PR has been created that supports the new feature #44421

**Pull request closed without merging**